File: leetcode/medium/paintHouse.cpp
class Solution {
public:
    int dp[2][3]{};
    int minCost(vector<vector<int>>& costs) {
        int N = costs.size();
        for (int i = 1; i <= N; ++i) {
            int t = i % 2;
            for (int c = 0; c < 3; ++c) {
                dp[t][c] = INT_MAX;
                for (int j = 0; j < 3; ++j) {
                    if (c == j) continue;
                    dp[t][c] = min(dp[t][c], dp[!t][j]);
                }
                dp[t][c] += costs[i-1][c];
            }
        }
        return *min_element(dp[N%2], dp[N%2] + 3);
    }
};

File: leetcode/medium/paint_house.py
class Solution:
    def minCost(self, costs: List[List[int]]) -> int:
        N = len(costs)
        @cache
        def dfs(i, c):
            if i == 0: return costs[0][c]
            ans = float('inf')
            for j in range(3):
                if c == j: continue
                ans = min(ans, dfs(i-1, j))
            return ans + costs[i][c]
        return min(dfs(N-1, c) for c in range(3))
